### DISCUSSION TITLE
is it a good idea to have a separate channel for each user?

### DISCUSSION THREAD

**Discussion by dhruv-nudge** [uNetworking/uWebSockets.js#1042] [Q&A] [answered]
I am working on a project where, I want to send a message to user based on some analytical events processing. 
There can be lot of users connected concurrently to our server at any point (upwards of 1 million).

so I wanted to know, is websocket the correct way to handle this? we only want to push data from server to client, and we will have APIs for ingesting the events data.

**Comment by uNetworkingAB** [comment, marked as answer]
Yes it's the ideal case. The pub/sub feature is very lightweight. But if you plan to send huge messages, that would be (more of) a problem. Small messages like, a text message or some notification or anything less than 1 KB is ideal use. Anything above 16 KB is probably going to jam up the server unless you scale it very well.

For 1 KB any shitty consumer grade computer can handle a million connections.

**Reply by uNetworkingAB** [reply]
If you need to send huge messages, I would split it up in two servers; one websocket server that sends a tiny notifications and then a HTTP server that is used to fetch the huge message after receiving the notification. IMO websockets should only be used for small message sending (but you can have absurd amounts of them)

**Reply by dhruv-nudge** [reply]
yes, I was thinking the same about having websockets server for only notifications and HTTP server for actual payload. Thanks for confirming that this approach will work.

Just one more question, if I have 2 channels for every client:
- one for user specific notifications (million channels with just one subscriber)
- one for broadcast information to all the users (so one channel can have upto million subscribers)

Assuming, I will always keep the payload < 1kb, both the above scenarios will work fine, right?

p.s. And like always thanks for wWebSockets.js

**Reply by uNetworkingAB** [reply]
Ok when I think more about it - yes it is okay to have 1 channel per connection BUT the same thing could be achieved without any channel at all, and instead just replace app.publish("per_connection_channel_13", message) with just connection13.send(message). Then you don't need a million of channels. You just need some hashmap from some ID to the socket (but that is kind of how it already is internally).

So to correct myself, I would not have a million channels just because, I would instead use ws.send directly so I only need the broadcast channel. But however you do it, it should not be a **problem** to have a million channels

**Reply by uNetworkingAB** [reply]
Another optimization you could think of is - if you can keep the notification message anonymous you don't even need SSL. And if non-SSL connection fails for some reason (let's say corporate proxy) it could fall back to SSL. If most of your users can use non-SSL (they will be able to) that can save a lot of cost.

**Reply by dhruv-nudge** [reply]
understood, thanks for the explanation, I will run some tests also later, to check what works best for us.

what if I use something like NGINX to manage SSL, and uWebSockets.js server only for websockets? do you think that will help?
